Feature - stdout live reporting

gecage952 opened the pull request:
Hi,
As part of our work at Oak Ridge National Lab, we've been using Galaxy for quite a while (some of us have attended GCC as well). We've also been doing some internal development on some features that users have requested here. One the most common ones was the ability to see live console output as jobs are running. This issue has been brought up before, for example https://github.com/galaxyproject/galaxy/issues/2332. But given, our users wanted it at the present, I took a stab at an implementation. My main goal was to minimize impact to the way Galaxy works now, so that there aren't any compatibility issues. I'll try to provide details about each part that was touched.

# Overview
The overall idea here was to add logic to allow the job manager to read the tool_stdout and tool_stderr files that are saved in the job directory, and return them as part of a status. The reason I put it into the status is because the UI already calls the status regularly form the JobInformation page, so I wouldn't have to make a new thread or anything. It also just kinda made sense to me that you might want it as part ofhte status of the job. To facilitate this, the api endpoint for getting job status was adjusted to allow parameters to select which part of the stdout/stderr (both work the same, so I'll just refer to stdout from here) that you want. There's `stdout_pos` which is the starting index in the stdout_file, and `stdout_length` which is how much of the stdout that you want (in chars). Because stdout could potentially be a relatively larger file, I didn't want to force people to read the whole file every time status is called. 

I then adjusted the UI for the job information view in a few different ways. First, I made the code blocks scrollable and set a max height for them. Then I moved the expand on click functionality over to only be on the expand icons, rather than the whole table row (if users would try and highlight a part of the stdout or click on the scroll bar to scroll, it would collapse the view). Lastly, I added an autoscroll feature that automatically scrolls the code blocks when the user is at the bottom of the stdout. If the user scrolls up, this is disabled. If they scroll back to the bottom, it starts again. 

The last thing I want to note is that as far as compatibility with job runners go, we almost exclusively use Pulsar for running jobs. *As is, this pr will only work for job runners that save their stdout to the job directory inside of the Galaxy.* Internally, we've added functionality for Pulsar to do this (the purpose of lib/galaxy/webapps/galaxy/api/job_files.py changes). I *did not* include those changes here, because that would require an additional pr to the Pulsar repository. Let me know if there's interest in seeing that however. It would also be nice to get some feedback on testing this.

I understand this is a pretty big change, and I imagine there are a lot of areas for improvement. Please let me know if this is something people are interested in helping with, or if this is a terrible way to try to do this, or whatever. It's worked for us so far internally, but I would love to have some feedback from here.


## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. The best way to test is to start a tool that will run for some time with stdout.
  2. Start the tool and go to the JobInformation page for the job.
  3. Expand the Tool stdout by clicking on the expand icon to the right. 

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

gecage952 committed:
Add option to load stdout and stderr when requesting job status

gecage952 committed:
Update api endpoint for jobs

gecage952 committed:
Add missing variables and import

gecage952 committed:
Add capability for UI to load live job stdout and stderr

gecage952 committed:
Merge pull request #1 from gecage952/dev

merge dev into branch

dannon committed:
Linting / format.

dannon committed:
Fix jest test

dannon committed:
Update API schema

mvdbeek commented on lib/galaxy/webapps/galaxy/api/jobs.py:
```diff
@@ -178,15 +178,31 @@ def show(
+        - stdout_position: The index of the character to begin reading stdout from
+        - stdout_length: How many characters of stdout to read
+        - stderr_position: The index of the character to begin reading stderr from
+        - stderr_length: How many characters of stderr to read
```
Can you add two separate endpoints job for the jobs' stdout and stderr ?

gecage952 committed:
Configure pulsar runner to read stdout from file if finish message stdout is none

gecage952 committed:
XMerge branch 'feature_stdout_live_reporting' of github.com:gecage952/galaxy into feature_stdout_live_reporting

gecage952 committed:
Merge branch 'dev' into feature_stdout_live_reporting

gecage952 committed:
Add new api endpoint to fetch job stdout and stderr

gecage952 committed:
Use stdout api endpoint in job info ui while job is running

gecage952 committed:
Merge upstream into branch

gecage952 committed:
Merge dev into branch and fix conflicts

gecage952 committed:
Update stdout reporting code to vue3

gecage952 committed:
Fix linting errors

gecage952 committed:
Fix more formatting and schema issues

gecage952 committed:
Merge branch 'dev' into feature_stdout_live_reporting

martenson committed:
format imports

gecage952 committed:
Fix issue where only one of stdout or stderr is present

gecage952 committed:
Merge branch 'feature_stdout_live_reporting' of github.com:gecage952/galaxy into feature_stdout_live_reporting

jmchilton commented on lib/galaxy/schema/jobs.py:
```diff
@@ -212,6 +212,12 @@ class JobOutput(Model):
 
 
+class JobConsoleOutput(Model):
+    state: Optional[JobState] = Field(None, title="Job State", description="The current job's state")
```
This is going to sounds weird but internally we track both a job_stderr/job_stdout and a tool_stderr/tool_stdout. Pulsar is worse at this than the other runners unfortunately. But I think the thing you're returning here is the tool_stdout - the output of the running tool. The job_stdout would be things like messages printed by SLURM and such. Can you update this description here to say Tool STDOUT and Tool STDERR. There are other places in comments where you refer to it as the job's stdout but I think it should be the job's tool stdout and such. It seems like there probably shouldn't be a distinction I imagine but we actually have disentangled a few problems this was and we shield the users from non-tool stuff they don't care about this way too.

gecage952 replied:
Gotcha.

jmchilton commented on lib/galaxy/managers/jobs.py:
```diff
@@ -273,6 +278,42 @@ def get_accessible_job(self, trans, decoded_job_id) -> Job:
+    ):
+        if job is None:
+            raise ObjectNotFound()
+
```
There are a bunch of reasons this might not be possible I think. Older Pulsar, remote execution outside of Pulsar, a Pulsar modality that doesn't allow this, runners that respect the location of working directory stuff in this fashion. It feels like the right thing to do here is to throw a ``galaxy.exceptions.ConfigDoesNotAllowException`` and then catch this on the client and not poll. Does that make sense to you?

Ideally I think we would make all this opt-in based on the job destination until we've hardened it and seen it work well in production without performance issues or confusing user experience. The job object has a ``job.job_destination`` YAML thing we could use to configure things per environment/destination.

gecage952 replied:
Yeah, that exception definitely makes sense. The compatibility just isn't going to be there for some stuff, I guess. Opt-in is a good idea as well.

gecage952 committed:
Fix docstrings that refer to tool stdout as job stdout

gecage952 committed:
Restrict tool stdout updating to properly configured job destinations

gecage952 committed:
update client api schema

gecage952 committed:
Fix merge conflicts

gecage952 committed:
Update client api schema

gecage952 committed:
Update client api schema

gecage952 committed:
Update client api schema

jmchilton committed:
Merge remote-tracking branch 'origin/dev' into feature_stdout_live_reporting

jmchilton committed:
Update JobInformation.test.js for live console output.

jmchilton committed:
Improved error handling for live console for tool execution.

jmchilton committed:
xfail the test for the freeze.

Got permission from Marius in exchange for pledge to fix it before the release.

jmchilton committed:
Merge remote-tracking branch 'origin/dev' into feature_stdout_live_reporting

jmchilton committed:
linting fix

jmchilton committed:
Ahhh... the set metadata file is JSON and cannot be appended.

So lets restrict this new append in the job files behavior to just append for the tool_stdout and tool_stderr files.
